### PR TITLE
MAINT: Move set_module from numpy.core to numpy._utils

### PR DESCRIPTION
**Pull request by seberg**
Note sure if others agree, this creates a `numpy._utils` module, because `numpy.core` already imports the C stuff.
The reason was that I wanted to use `set_module` when moving exceptions (since they currently have that).  But `set_module` being defined in `overrides` needs C to be imported.  

So I thought, maybe we actually need a place to put things that don't need depend on anything else to avoid circular imports...

If nobody else likes the idea, I am happy to close and just use `object.__module__ = "numpy"`, or put it elsewhere (but I didn't have a good idea for that).

### CODE REVIEW

**Comment by seberg** [comment]
Note that I have not changed imports that are using `from ...overrides import something, set_module`, since they need the rest anyway.

**Comment by mattip** [comment]
Maybe make this `numpy.lib._utils`?

**Comment by seberg** [comment]
Yeah, I wanted to put it somewhere deeper, and I don't quite love this either (so happy for other ideas even duplicating the helper).
The problem is, I also want this to be available before any C import happens...  `lib` pulls in the public submodules, which pull in all of NumPy (circular imports).

**Comment by mattip** [comment]
I guess `numpy._utils` makes sense. Perhaps we should move `numpy/compat/_inspect.py` there as well.

**Comment by mattip** [comment]
Ahh, could we have a `numpy._utils` (maybe a different name?) directory, and in that have a `set_module.py` file? Then in the `__init__.py` it could say something about "A place for code that does not import from the rest of NumPy, to reduce circular dependencies"

**Comment by seberg** [comment]
Yeah, a submodule/folder make sense, how about being super clear, like `_pydefs`?

`_globals` and `_inspect` could go there as well, then, yeah.  Maybe more?  I am not sure I love it, but a place to reduce circular imports seems like it likely makes sense anyway.

**Comment by mattip** [comment]
> how about being super clear, like `_pydefs`

I am bad with names. We are looking for a place to put things with no dependencies. Do we want to definitely restrict ourselves to pure-python modules or can we also put some simple no-dependency c-extension module in the future? I think I prefer something in the direction of `utils` or `basic`

**Comment by seberg** [comment]
Hmmm, yeah, i don't the name `_pydefs` either anyway.  `_utils` seems fine to me, it is at least a place where you would look for these things?

Created a module, left `set_module` in the `__init__` for the moment.  Moved `_inspect` and `_pep440`.  `_inspect` is actuall imported to `np.compat`, but maybe it makes sense to move it anyway?

File: numpy/compat/__init__.py [hunk 1]
```diff
@@ -8,9 +8,10 @@
   * we may only need a small subset of the copied library/module
 
 """
-from . import _inspect
+
+from .._utils import _inspect
```
Comment (mattip) [inline numpy/compat/__init__.py:12]: Is this line required?

**Review by mattip — APPROVED** [review]
LGTM.

**Comment by mattip** [comment]
I will put this in soon if there are no further comments

**Comment by mattip** [comment]
Thanks @seberg

**Pull request merged**